HPCC-21762 Spray does not set file type and incorrectly sets format (for CSV)

AttilaVamos opened the pull request:
Add code to determine and store correct info into @kind and @format
attributes.

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

<!-- Thank you for submitting a pull request to the HPCC project
 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

Smoketest:
- [x] Send notifications about my Pull Request position in Smoketest queue.

## Testing:
<!-- Please describe how this change has been tested.-->
Tested manually.
<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

HPCCSmoketest commented:
Process of PR-12380, label: HPCC-21762-fix-7.2.x starts now.
The reason of this test is: Did not build yet.
Commit ID: 097fde4ace3d02f4b83a1449f8f8dad6d11f05d0
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 097fde4ace3d02f4b83a1449f8f8dad6d11f05d0
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |111 |111 |0 |0 |0 |35 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |25 sec  (00:00:25) |
|setup (thor) |11 |11 |0 |41 sec  (00:00:41) |
|setup (roxie) |11 |11 |0 |18 sec  (00:00:18) |
|test (hthor) |823 |823 |0 |148 sec  (00:02:28) |
|test (thor) |748 |748 |0 |622 sec  (00:10:22) |
|test (roxie) |897 |897 |0 |207 sec  (00:03:27) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   18 sec (00:00:18) |  211 sec (00:03:31) |    0 sec (00:00:00) |    3 sec (00:00:03) |   18 sec (00:00:18) | 1272 sec (00:21:12) |   19 sec (00:00:19) | 1541 sec (00:25:41) |

AttilaVamos commented:
@jakesmith please review.

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -3224,8 +3224,68 @@ void FileSprayer::updateTargetProperties()
+
+                StringBuffer targetFormat = "utf8n";
+
+                switch (srcFormat.type)
```
Is this right? 
Shouldn't tgtFormat.type be used here?

The source format is not necessarily going to be same as target format.

AttilaVamos replied:
Changed

jakesmith commented:
@ghalliday - could you take a look at this too?

ghalliday commented on dali/ft/filecopy.cpp:
```diff
@@ -3224,8 +3224,68 @@ void FileSprayer::updateTargetProperties()
+                    targetKind = "csv";
+                    break;
+                case FFTblocked:
+                    targetKind = "blocked";
```
You could actually set the record size to EFX_BLOCK_SIZE in this case.

AttilaVamos replied:
If the "blocked" is not a published file kind then what kind of file should be set along with the suggested record size?

jakesmith replied:
I think it should be "flat"

AttilaVamos replied:
Changed

ghalliday commented on dali/ft/filecopy.cpp:
```diff
@@ -3224,8 +3224,68 @@ void FileSprayer::updateTargetProperties()
+                    targetKind ="fixed";
+                    break;
+                case FFTvariable:
+                    targetKind = "csv";
```
I've double checked - they appear to be 4 byte prefixed.

AttilaVamos replied:
What is the kind of "FTTvariable" if it isn't csv? In other hand which FileFormatType value identifies the "csv" file?

jakesmith replied:
Should be "flat" I think.

AttilaVamos replied:
Changed

ghalliday commented:
@AttilaVamos Jake's comments look correct

AttilaVamos commented on dali/ft/filecopy.cpp:
```diff
@@ -3224,8 +3224,68 @@ void FileSprayer::updateTargetProperties()
+                switch (srcFormat.type)
+                {
+                case FFTfixed:
+                    targetKind ="fixed";
```
What file kind should be set for "FFTFixed"?

jakesmith replied:
"flat"

AttilaVamos replied:
Changed

AttilaVamos commented on dali/ft/filecopy.cpp:
```diff
@@ -3224,8 +3224,68 @@ void FileSprayer::updateTargetProperties()
+                }
 
+                curProps.setProp("@kind", targetKind.str());
+                curProps.setProp("@format", targetFormat.str());
```
Changed.
I think there is no difference between FP* and FA* but two different development branches in the past. I think it needs a separate JIRA to remove this kind of duplication.

jakesmith replied:
sure, but should use one of the definition not sure which, instead of const strings above.

ghalliday replied:
I can't remember the difference.  I think they were used for  two logically different structures, but no need for them to differ

AttilaVamos commented on dali/ft/filecopy.cpp:
```diff
@@ -3224,8 +3224,68 @@ void FileSprayer::updateTargetProperties()
+                else if (srcFormat.markup == FMTjson)
+                    targetKind = "json";
+
+                StringBuffer targetFormat = "utf8n";
```
Right.

AttilaVamos commented:
@jakesmith @ghalliday I have some questions, need some clarifications.

jakesmith commented:
@ghalliday - can you see my latest replies to Attila's questions and correct/confirm?

ghalliday reviewed (commented):
@jakesmith I agree with your replies.  @AttilaVamos

AttilaVamos commented on dali/ft/filecopy.cpp:
```diff
@@ -3224,8 +3224,68 @@ void FileSprayer::updateTargetProperties()
+                curProps.setProp("@kind", srcAttr->queryProp("@kind"));
+            else
+            {
+                StringBuffer targetKind = "csv";
```
Changed

HPCCSmoketest commented:
Process of PR-12380, label: HPCC-21762-fix-7.2.x starts now.
The reason of this test is: New commit.
Commit ID: 382229d5cb5c8126b5d3cbd6cc5db97106a73908
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 382229d5cb5c8126b5d3cbd6cc5db97106a73908
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |111 |111 |0 |0 |0 |38 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |23 sec  (00:00:23) |
|setup (thor) |11 |11 |0 |45 sec  (00:00:45) |
|setup (roxie) |11 |11 |0 |21 sec  (00:00:21) |
|test (hthor) |825 |825 |0 |200 sec  (00:03:20) |
|test (thor) |750 |750 |0 |594 sec  (00:09:54) |
|test (roxie) |901 |901 |0 |202 sec  (00:03:22) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   24 sec (00:00:24) |  213 sec (00:03:33) |    0 sec (00:00:00) |    2 sec (00:00:02) |   19 sec (00:00:19) | 1306 sec (00:21:46) |   20 sec (00:00:20) | 1584 sec (00:26:24) |

AttilaVamos commented:
@jakesmith please re-review

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -3224,8 +3224,70 @@ void FileSprayer::updateTargetProperties()
+                }
 
+                if (! streq(targetKind, "unknown"))
+                    curProps.setProp(FPkind, targetKind);
```
better to set targetKind to nullptr and avoid the streq to "unknown", i.e. make them:
```
if (targetKind)
    curProps.setProp(FPkind, targetKind);
```

AttilaVamos replied:
Changed

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -3224,8 +3224,70 @@ void FileSprayer::updateTargetProperties()
+                    curProps.setProp(FPkind, targetKind);
+
+                if (! streq(targetFormat, "unknown"))
+                    curProps.setProp(FPformat, targetFormat);
```
as per targetKind

AttilaVamos replied:
Changed

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -3224,8 +3224,70 @@ void FileSprayer::updateTargetProperties()
             // Keep source kind
-            curProps.setProp("@kind", srcAttr->queryProp("@kind"));
+            if (srcAttr->hasProp("@kind"))
+                curProps.setProp("@kind", srcAttr->queryProp("@kind"));
```
should you also check srcAttr->hasProp("@format") in similar way?

jakesmith replied:
@ghalliday - is it possible for source and target to be different kinds? e.g. spray csv to flat file?

AttilaVamos replied:
@ghalliday is there any conclusion related this question?

ghalliday replied:
It is possible for the source to be utf16 and the target to be utf8.  There are situations where files are sprayed as csv, but actually ready as xml, but I don't think the basic csv/xml/flat format ever changes.

jakesmith replied:
ok, then I think that possibility is not an issue for this JIRA/PR - perhaps it _should_ be possible to spray and change target published format, but not something that's currently possible.

jakesmith replied:
> https://github.com/hpcc-systems/HPCC-Platform/pull/12380/files#r278079968

@AttilaVamos  Not sure if this has/needs to be addressed ?

i.e. shouldn't above code look something like:
```
            if (srcAttr->hasProp("@kind"))
            {
                curProps.setProp("@kind", srcAttr->queryProp("@kind"));
                if (srcAttr->hasProp(FPformat))
                    curProps.setProp(FPformat, srcAttr->queryProp(FPformat));
            }
```

jakesmith commented:
@AttilaVamos - a couple of comments and a question re. spraying between types.

HPCCSmoketest commented:
Process of PR-12380, label: HPCC-21762-fix-7.2.x starts now.
The reason of this test is: New commit.
Commit ID: dd1e68c93c14a6ea0c8dd9a880e638efcd73b612
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: dd1e68c93c14a6ea0c8dd9a880e638efcd73b612
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |113 |113 |0 |0 |0 |34 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |23 sec  (00:00:23) |
|setup (thor) |11 |11 |0 |44 sec  (00:00:44) |
|setup (roxie) |11 |11 |0 |27 sec  (00:00:27) |
|test (hthor) |828 |828 |0 |192 sec  (00:03:12) |
|test (thor) |753 |753 |0 |610 sec  (00:10:10) |
|test (roxie) |904 |904 |0 |231 sec  (00:03:51) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   28 sec (00:00:28) |  213 sec (00:03:33) |    0 sec (00:00:00) |    3 sec (00:00:03) |   18 sec (00:00:18) | 1344 sec (00:22:24) |   20 sec (00:00:20) | 1626 sec (00:27:06) |

ghalliday reviewed (commented):
@jakesmith does that answer your question?

HPCCSmoketest commented:
Process of PR-12380, label: HPCC-21762-fix-7.2.x starts now.
The reason of this test is: New commit.
Commit ID: 895f0e7f2c1ee763e4a9ad05b6735a30c3b9d454
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 895f0e7f2c1ee763e4a9ad05b6735a30c3b9d454
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |113 |113 |0 |0 |0 |35 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |24 sec  (00:00:24) |
|setup (thor) |11 |11 |0 |45 sec  (00:00:45) |
|setup (roxie) |11 |11 |0 |24 sec  (00:00:24) |
|test (hthor) |829 |829 |0 |156 sec  (00:02:36) |
|test (thor) |754 |754 |0 |608 sec  (00:10:08) |
|test (roxie) |905 |905 |0 |409 sec  (00:06:49) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   34 sec (00:00:34) |  209 sec (00:03:29) |    0 sec (00:00:00) |    2 sec (00:00:02) |   18 sec (00:00:18) | 1507 sec (00:25:07) |   21 sec (00:00:21) | 1791 sec (00:29:51) |

AttilaVamos committed:
HPCC-21762 Spray does not set file type and incorrectly sets format (for CSV)

Add code to determine and store correct info into @kind and @format
attributes.

Tested manually.

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

HPCCSmoketest commented:
Process of PR-12380, label: HPCC-21762-fix-7.2.x starts now.
The reason of this test is: New commit.
Commit ID: 175f048ca834eee5f7505764e6e318f67e60f087
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 175f048ca834eee5f7505764e6e318f67e60f087
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |113 |113 |0 |0 |0 |34 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |28 sec  (00:00:28) |
|setup (thor) |11 |11 |0 |44 sec  (00:00:44) |
|setup (roxie) |11 |11 |0 |17 sec  (00:00:17) |
|test (hthor) |829 |829 |0 |192 sec  (00:03:12) |
|test (thor) |754 |754 |0 |598 sec  (00:09:58) |
|test (roxie) |905 |905 |0 |223 sec  (00:03:43) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   31 sec (00:00:31) |  209 sec (00:03:29) |    0 sec (00:00:00) |    3 sec (00:00:03) |   18 sec (00:00:18) | 1319 sec (00:21:59) |   19 sec (00:00:19) | 1599 sec (00:26:39) |

AttilaVamos commented:
@jakesmith squashed.

jakesmith commented:
@AttilaVamos - I think this is not suitable for point build, can you rebase to master (so for 7.4) ?

AttilaVamos commented:
This PR replaced by the re-targeted  https://github.com/hpcc-systems/HPCC-Platform/pull/12564.

jakesmith commented:
@AttilaVamos -  you could have retargeted this PR without closing it/opening another.
